### DISCUSSION TITLE
Writing plugin for TS: different results in execution in Rust context or in JS context

### DISCUSSION THREAD

**Discussion by amanteaux** [swc-project/swc#9130] [Q&A] [answered]
Hello,

I want to build a plugin that adds metadata on TS classes based on the constructor types. This is for dependency injection, my final goal is to replace the TS transformer https://github.com/wessberg/DI-compiler

I am using the hooks `fn visit_mut_class_decl(&mut self, class_decl: &mut ClassDecl)` and `fn visit_mut_class_expr(&mut self, class_expr: &mut ClassExpr)` to get class data.
I am the using this code to extract constructor information:
```rust
        for member in &class.body {
            if let ClassMember::Constructor(constructor) = member {
                println!("ClassMember::Constructor");
                for param in &constructor.params {
// [...]
```

Everything works perfectly when executed in a Rust environment: I correctly get params as a `TsParamProp`, so I am able to recover the param type.
However, when I execute the plugin from JS, I get params as `Param`: so the type information seems erased, as if I was getting the JS version of the TS file I want to compile. 

Here is my JSC JS parser configuration:
```javascript
      parser: {
        syntax: "typescript",
        tsx: true,
        decorators: true,
      }
```

Here are some debug information I get, from Rust execution:
```
TsParamProp { span: 180..217#0, decorators: [], accessibility: Some(Private), is_override: false, readonly: true, param: Ident(BindingIdent { id: Ident { span: 197..206#0, sym: "sampleApi", optional: false }, type_ann: Some(TsTypeAnn { span: 206..217#0, type_ann: TsTypeRef(TsTypeRef { span: 208..217#0, type_name: Ident(Ident { span: 208..217#0, sym: "SampleApi", optional: false }), type_params: None }) }) }) }
```
From JS execution (on the same TS file where types seems to be erased):
```
Param { span: 181..218#0, decorators: [], pat: Ident(BindingIdent { id: Ident { span: 198..207#3, sym: "sampleApi", optional: false }, type_ann: None }) }
```

I use in JS the `@swc/core` in version `1.6.6` whereas in Rust I am using `swc_core 0.95.10` and `swc_ecma_parser 0.146.7`.

I feel like I am missing a simple option that would enable me to correctly get the `TsParamProp` type. But I cannot find it!
Any ideas?

I can provide all my code if necessary.
Thank you for your help!

**Comment by kdy1** [comment, marked as answer]
TypeScript/decorators are handled before running Wasm plugins. I'll document it.

**Reply by kdy1** [reply]
See https://github.com/swc-project/swc/issues/9132

**Reply by amanteaux** [reply]
Thank you very much for your quick answer! Having TypeScript transformers would enable to completely remove TSC compilation from our build tools, it would be really awesome!

If I can help in any way to implement https://github.com/swc-project/swc/issues/9132, please tell me.

**Reply by amanteaux** [reply]
Hello @kdy1,

I wanted to thank you for adding Wasm TS plugins to SWC. We have been able to port our dependency injection TS plugin to SWC: https://github.com/Coreoz/SWC-class-decorator-plugin

Our projects can now start and build at least 2 or 3 times faster (the bigger the project, the bigger the gain is).

Cheers,
Aurélien
